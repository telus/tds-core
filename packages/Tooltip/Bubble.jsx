import React from 'react'
import PropTypes from 'prop-types'

import joinClassNames from '../../shared/utils/joinClassNames'

import Text from '../Text/Text'
import Box from '../Box/Box'

import styles from './Tooltip.modules.scss'
import displayStyles from '../../shared/styles/Display.modules.scss'

const Bubble = ({ id, direction, width, open, children }) => {
  const classes = joinClassNames(styles[direction], styles[width], !open && displayStyles.hide)

  return (
    <Box
      vertical={2}
      horizontal={3}
      dangerouslyAddClassName={classes}
      id={id}
      role="tooltip"
      aria-live="polite"
      aria-hidden={open ? 'false' : 'true'}
      data-testid="bubble"
    >
      <Text size="small">{children}</Text>
    </Box>
  )
}

Bubble.propTypes = {
  id: PropTypes.string.isRequired,
  direction: PropTypes.oneOf(['left', 'right']).isRequired,
  width: PropTypes.oneOf(['full', 'half', 'quarter']).isRequired,
  open: PropTypes.bool.isRequired,
  children: PropTypes.node.isRequired,
}

export default Bubble
