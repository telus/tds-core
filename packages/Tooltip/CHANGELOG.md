# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [4.2.16](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.2.15...@tds/core-tooltip@4.2.16) (2020-07-22)

**Note:** Version bump only for package @tds/core-tooltip





## [4.2.15](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.2.14...@tds/core-tooltip@4.2.15) (2020-04-14)

**Note:** Version bump only for package @tds/core-tooltip





## [4.2.14](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.2.13...@tds/core-tooltip@4.2.14) (2020-03-17)

**Note:** Version bump only for package @tds/core-tooltip





## [4.2.13](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.2.12...@tds/core-tooltip@4.2.13) (2020-03-11)

**Note:** Version bump only for package @tds/core-tooltip





## [4.2.12](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.2.11...@tds/core-tooltip@4.2.12) (2020-03-10)

**Note:** Version bump only for package @tds/core-tooltip





## [4.2.11](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.2.10...@tds/core-tooltip@4.2.11) (2020-03-09)

**Note:** Version bump only for package @tds/core-tooltip





## [4.2.10](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.2.9...@tds/core-tooltip@4.2.10) (2020-02-24)

**Note:** Version bump only for package @tds/core-tooltip





## [4.2.9](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.2.8...@tds/core-tooltip@4.2.9) (2020-02-03)

**Note:** Version bump only for package @tds/core-tooltip





## [4.2.8](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.2.7...@tds/core-tooltip@4.2.8) (2020-01-27)

**Note:** Version bump only for package @tds/core-tooltip





## [4.2.7](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.2.6...@tds/core-tooltip@4.2.7) (2020-01-23)

**Note:** Version bump only for package @tds/core-tooltip





## [4.2.6](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.2.5...@tds/core-tooltip@4.2.6) (2020-01-22)

**Note:** Version bump only for package @tds/core-tooltip





## [4.2.5](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.2.4...@tds/core-tooltip@4.2.5) (2019-11-20)

**Note:** Version bump only for package @tds/core-tooltip





## [4.2.4](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.2.3...@tds/core-tooltip@4.2.4) (2019-11-18)

**Note:** Version bump only for package @tds/core-tooltip





## [4.2.3](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.2.2...@tds/core-tooltip@4.2.3) (2019-10-09)

**Note:** Version bump only for package @tds/core-tooltip





## [4.2.2](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.2.1...@tds/core-tooltip@4.2.2) (2019-10-04)

**Note:** Version bump only for package @tds/core-tooltip





## [4.2.1](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.2.0...@tds/core-tooltip@4.2.1) (2019-09-26)

**Note:** Version bump only for package @tds/core-tooltip





# [4.2.0](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.1.4...@tds/core-tooltip@4.2.0) (2019-09-18)


### Features

* **core-tooltip:** allow forwarding refs ([1fe1474](https://github.com/telusdigital/tds/commit/1fe1474))





## [4.1.4](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.1.3...@tds/core-tooltip@4.1.4) (2019-09-12)

**Note:** Version bump only for package @tds/core-tooltip





## [4.1.3](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.1.2...@tds/core-tooltip@4.1.3) (2019-09-06)

**Note:** Version bump only for package @tds/core-tooltip





## [4.1.2](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.1.1...@tds/core-tooltip@4.1.2) (2019-08-26)

**Note:** Version bump only for package @tds/core-tooltip





## [4.1.1](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.1.0...@tds/core-tooltip@4.1.1) (2019-08-22)

**Note:** Version bump only for package @tds/core-tooltip





# [4.1.0](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.0.8...@tds/core-tooltip@4.1.0) (2019-08-20)


### Features

* **core-tooltip:** generate unique id using @tds/util-helpers uniqueId ([bc8c87d](https://github.com/telusdigital/tds/commit/bc8c87d)), closes [#1223](https://github.com/telusdigital/tds/issues/1223)





## [4.0.8](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.0.7...@tds/core-tooltip@4.0.8) (2019-08-09)

**Note:** Version bump only for package @tds/core-tooltip





## [4.0.7](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.0.6...@tds/core-tooltip@4.0.7) (2019-08-01)

**Note:** Version bump only for package @tds/core-tooltip





## [4.0.6](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.0.5...@tds/core-tooltip@4.0.6) (2019-07-31)

**Note:** Version bump only for package @tds/core-tooltip





## [4.0.5](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.0.4...@tds/core-tooltip@4.0.5) (2019-07-30)

**Note:** Version bump only for package @tds/core-tooltip





## [4.0.4](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.0.3...@tds/core-tooltip@4.0.4) (2019-07-25)

**Note:** Version bump only for package @tds/core-tooltip





## [4.0.3](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.0.2...@tds/core-tooltip@4.0.3) (2019-07-17)


### Bug Fixes

* **core-tooltip:** add button type for tooltip button ([584639d](https://github.com/telusdigital/tds/commit/584639d)), closes [#1187](https://github.com/telusdigital/tds/issues/1187)





## [4.0.2](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.0.1...@tds/core-tooltip@4.0.2) (2019-07-17)

**Note:** Version bump only for package @tds/core-tooltip





## [4.0.1](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@4.0.0...@tds/core-tooltip@4.0.1) (2019-07-11)

**Note:** Version bump only for package @tds/core-tooltip





# [4.0.0](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@3.0.7...@tds/core-tooltip@4.0.0) (2019-07-10)


### Features

* **core-tooltip:** add copy prop and documentation ([c7bad06](https://github.com/telusdigital/tds/commit/c7bad06))


### BREAKING CHANGES

* **core-tooltip:** new copy prop is required





## [3.0.7](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@3.0.6...@tds/core-tooltip@3.0.7) (2019-06-26)

**Note:** Version bump only for package @tds/core-tooltip





## [3.0.6](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@3.0.5...@tds/core-tooltip@3.0.6) (2019-06-25)

**Note:** Version bump only for package @tds/core-tooltip





## [3.0.5](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@3.0.4...@tds/core-tooltip@3.0.5) (2019-06-04)

**Note:** Version bump only for package @tds/core-tooltip

## [3.0.4](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@3.0.3...@tds/core-tooltip@3.0.4) (2019-05-07)

**Note:** Version bump only for package @tds/core-tooltip

## [3.0.3](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@3.0.2...@tds/core-tooltip@3.0.3) (2019-05-01)

**Note:** Version bump only for package @tds/core-tooltip

## [3.0.2](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@3.0.1...@tds/core-tooltip@3.0.2) (2019-05-01)

**Note:** Version bump only for package @tds/core-tooltip

## [3.0.1](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@3.0.0...@tds/core-tooltip@3.0.1) (2019-04-23)

**Note:** Version bump only for package @tds/core-tooltip

# [3.0.0](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@2.0.13...@tds/core-tooltip@3.0.0) (2019-04-22)

### Features

- **core-tooltip:** convert to styled components ([4fbedd3](https://github.com/telusdigital/tds/commit/4fbedd3))

### BREAKING CHANGES

- **core-tooltip:** now uses styled components

## [2.0.13](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@2.0.12...@tds/core-tooltip@2.0.13) (2019-04-11)

### Bug Fixes

- **core-tooltip:** use className prop for Box ([cc7315a](https://github.com/telusdigital/tds/commit/cc7315a))

## [2.0.12](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@2.0.11...@tds/core-tooltip@2.0.12) (2019-03-28)

**Note:** Version bump only for package @tds/core-tooltip

## [2.0.11](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@2.0.10...@tds/core-tooltip@2.0.11) (2019-03-20)

**Note:** Version bump only for package @tds/core-tooltip

## [2.0.10](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@2.0.9...@tds/core-tooltip@2.0.10) (2019-03-20)

**Note:** Version bump only for package @tds/core-tooltip

## [2.0.9](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@2.0.8...@tds/core-tooltip@2.0.9) (2019-03-18)

**Note:** Version bump only for package @tds/core-tooltip

## [2.0.8](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@2.0.7...@tds/core-tooltip@2.0.8) (2019-03-18)

**Note:** Version bump only for package @tds/core-tooltip

## [2.0.7](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@2.0.6...@tds/core-tooltip@2.0.7) (2019-03-11)

**Note:** Version bump only for package @tds/core-tooltip

## [2.0.6](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@2.0.5...@tds/core-tooltip@2.0.6) (2019-03-05)

**Note:** Version bump only for package @tds/core-tooltip

## [2.0.5](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@2.0.4...@tds/core-tooltip@2.0.5) (2019-02-06)

**Note:** Version bump only for package @tds/core-tooltip

<a name="2.0.4"></a>

## [2.0.4](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@2.0.3...@tds/core-tooltip@2.0.4) (2018-12-19)

### Bug Fixes

- **core-tooltip:** prevent multiple tooltips from being open on one page ([8bc243d](https://github.com/telusdigital/tds/commit/8bc243d))

<a name="2.0.3"></a>

## [2.0.3](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@2.0.2...@tds/core-tooltip@2.0.3) (2018-12-06)

**Note:** Version bump only for package @tds/core-tooltip

<a name="2.0.2"></a>

## [2.0.2](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@2.0.1...@tds/core-tooltip@2.0.2) (2018-07-23)

**Note:** Version bump only for package @tds/core-tooltip

<a name="2.0.1"></a>

## [2.0.1](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@2.0.0...@tds/core-tooltip@2.0.1) (2018-07-13)

### Bug Fixes

- **core-tooltip:** add fix for line-height inherited from core-text ([262ab8b](https://github.com/telusdigital/tds/commit/262ab8b))

<a name="2.0.0"></a>

# [2.0.0](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@1.1.1...@tds/core-tooltip@2.0.0) (2018-07-09)

### Features

- **core-tooltip:** change direction prop default to auto ([9835ce0](https://github.com/telusdigital/tds/commit/9835ce0))

### BREAKING CHANGES

- **core-tooltip:** direction prop is set to auto by default instead of right

<a name="1.1.1"></a>

## [1.1.1](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@1.1.0...@tds/core-tooltip@1.1.1) (2018-07-09)

### Bug Fixes

- **core-tooltip:** revert default value for the direction prop ([432da53](https://github.com/telusdigital/tds/commit/432da53))

<a name="1.1.0"></a>

# [1.1.0](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@1.0.4...@tds/core-tooltip@1.1.0) (2018-07-09)

### Bug Fixes

- **core-tooltip:** adjust bubble size when facing right ([c8cba07](https://github.com/telusdigital/tds/commit/c8cba07))
- **core-tooltip:** remove width prop from bubble, use SASS mq ([a7f2119](https://github.com/telusdigital/tds/commit/a7f2119))

### Features

- **core-tooltip:** add ability to be standalone, automate bubble side ([#666](https://github.com/telusdigital/tds/issues/666)) ([2bfa0bb](https://github.com/telusdigital/tds/commit/2bfa0bb))
- **core-tooltip:** add auto as a direction wip ([3288333](https://github.com/telusdigital/tds/commit/3288333))
- **core-tooltip:** add bubble styles ([9921e87](https://github.com/telusdigital/tds/commit/9921e87))

<a name="1.0.4"></a>

## [1.0.4](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@1.0.3...@tds/core-tooltip@1.0.4) (2018-06-18)

### Bug Fixes

- **core-tooltip:** enhance ability for screen readers to see text ([e1d5c7f](https://github.com/telusdigital/tds/commit/e1d5c7f))

<a name="1.0.3"></a>

## [1.0.3](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@1.0.2...@tds/core-tooltip@1.0.3) (2018-06-14)

**Note:** Version bump only for package @tds/core-tooltip

<a name="1.0.2"></a>

## [1.0.2](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@1.0.1...@tds/core-tooltip@1.0.2) (2018-05-14)

### Bug Fixes

- **core-tooltip:** remove unused responsive import ([0af92d5](https://github.com/telusdigital/tds/commit/0af92d5))

<a name="1.0.1"></a>

## [1.0.1](https://github.com/telusdigital/tds/compare/@tds/core-tooltip@1.0.0...@tds/core-tooltip@1.0.1) (2018-04-02)

### Bug Fixes

- **misc:** remove styleguidist errors for markdownheading ([dac25cc](https://github.com/telusdigital/tds/commit/dac25cc))
