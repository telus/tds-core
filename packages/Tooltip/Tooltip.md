Tooltips provide a descriptive and detailed explanation or instructions to help the user fill in the form field. You may
use one when the information is useful only to a small percentage of users (ie. tech savvy people wouldn't need this info).
Tooltips may also be useful when vertical space is an issue.

Use the `direction` prop to open the tooltip to the left or right of the trigger, so that it does not block other
content on the page. Be careful not to let the Tooltip extend beyond the edge of the page if opening to the right though.

On small screen sizes Tooltips will always open to the left to ensure they will be visible.

```
<Input
  label="Enter your device's IMEI number"
  type="number"
  tooltip={
    <Tooltip direction="left">
      You can find your device's IMEI by typing *#06# on its dialpad. Please enter the complete 15-digit number, without spaces or characters.
    </Tooltip>
  }
/>
```
