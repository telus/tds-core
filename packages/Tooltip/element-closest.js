// DO NOT MODIFY THIS FILE. IT IS COPIED DIRECTLY FROM AN NPM PACKAGE

// This was copied from the element-closest polyfill: github.com/jonathantneal/closest
// This version does not attempt to polyfill `Element` because window.Element is not available server side. This version
// is less invasive.

const matches = (element, selector) => {
  const elements = (element.document || element.ownerDocument).querySelectorAll(selector)
  let index = 0

  while (elements[index] && elements[index] !== element) {
    index += 1
  }

  return Boolean(elements[index])
}

const closest = (element, selector) => {
  let currentElement = element

  while (currentElement && currentElement.nodeType === 1) {
    if (matches(currentElement, selector)) {
      return currentElement
    }

    currentElement = currentElement.parentNode
  }

  return null
}

export default closest
