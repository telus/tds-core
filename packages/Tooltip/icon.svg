<?xml version="1.0" encoding="UTF-8"?>
<svg width="24px" height="24px" viewBox="0 0 22 22">
  <path d="M10.9995305,0 C17.0650958,0 22,4.93396508 22,10.9995305 C22,17.0650958 17.0650958,22 10.9995305,22 C4.93396508,22 0,17.0650958 0,10.9995305 C0,4.93396508 4.93396508,0 10.9995305,0 Z M10.9995305,14.4580015 C10.3543774,14.4580015 9.83036667,14.9820122 9.83036667,15.6281043 C9.83036667,16.2741964 10.3543774,16.7972681 10.9995305,16.7972681 C11.6456226,16.7972681 12.1696333,16.2741964 12.1696333,15.6281043 C12.1696333,14.9820122 11.6456226,14.4580015 10.9995305,14.4580015 Z M10.9554873,5.45271695 C10.4192684,5.45271695 9.93188202,5.54380843 9.49802365,5.72505229 C9.0622871,5.90535707 8.69228668,6.1560934 8.38426602,6.47820037 C8.07718445,6.80124643 7.83677808,7.18157681 7.66680326,7.62200879 C7.49401118,8.06150169 7.4038588,8.5441926 7.39446792,9.07289879 L7.39446792,9.07289879 L9.54967345,9.07289879 L9.55499984,8.87817022 C9.56569663,8.68664041 9.5926367,8.50475093 9.63700858,8.33289794 C9.695232,8.10282153 9.78538439,7.90279592 9.90840483,7.73188202 C10.0304862,7.56096811 10.1863747,7.42386136 10.3770094,7.32150083 C10.5676442,7.21820122 10.7949033,7.1674905 11.0578478,7.1674905 C11.4485081,7.1674905 11.7546506,7.27548555 11.9743971,7.48959747 C12.1941435,7.70558757 12.3049558,8.03708541 12.3049558,8.48596918 C12.3134076,8.75079182 12.2673923,8.97053827 12.1650318,9.14614761 C12.0626713,9.32175695 11.9274427,9.48234089 11.7612242,9.62977761 C11.5959448,9.77627524 11.414701,9.92277287 11.2193708,10.0683314 C11.0231015,10.2157681 10.8390404,10.3885602 10.662492,10.5895249 C10.4868827,10.7895505 10.3328723,11.0318351 10.200461,11.3145004 C10.0699279,11.5981048 9.98822726,11.9502625 9.95911555,12.3700346 L9.95911555,12.3700346 L9.95911555,13.030213 L11.9368336,13.030213 L11.9368336,12.4723951 L11.9672525,12.3024578 C12.0168062,12.0836881 12.1023008,11.8961711 12.2232552,11.7399069 C12.3753874,11.5445768 12.5472404,11.3708456 12.7425705,11.2196525 C12.9388398,11.0684595 13.1463781,10.9172664 13.3670636,10.7660733 C13.5868101,10.6139412 13.7868357,10.4308191 13.9671405,10.2157681 C14.1474453,10.0007171 14.3005165,9.74246809 14.4207197,9.43914287 C14.5437401,9.13581765 14.6047808,8.75079182 14.6047808,8.28124813 C14.6047808,7.99764374 14.5437401,7.69244035 14.4207197,7.36563794 C14.3005165,7.03789644 14.0967345,6.73363213 13.8131301,6.44908866 C13.5304648,6.16548427 13.155769,5.92883425 12.6927989,5.73819951 C12.2288897,5.54850386 11.6494728,5.45271695 10.9554873,5.45271695 Z" />
</svg>
