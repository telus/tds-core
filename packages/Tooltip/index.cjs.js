const Tooltip = require('./dist/index.cjs')

module.exports = Tooltip
