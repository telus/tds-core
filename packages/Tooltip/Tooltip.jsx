import React from 'react'
import PropTypes from 'prop-types'

import Responsive from '@tds/core-responsive'
import StandaloneIcon from '@tds/core-standalone-icon'

import safeRest from '../../shared/utils/safeRest'
import generateId from '../../shared/utils/generateId'
import closest from './element-closest'

import Bubble from './Bubble'

import iconWrapperStyles from '../../shared/styles/IconWrapper.modules.scss'

const getTriggerA11yText = connectedFieldLabel => {
  if (!connectedFieldLabel) {
    return 'Reveal additional information.'
  }

  return `Reveal additional information about ${connectedFieldLabel}.`
}

const getIds = connectedFieldLabel => {
  const id = generateId(connectedFieldLabel, 'unknown-field')

  return {
    bubbleId: id.postfix('tooltip'),
    triggerId: id.postfix('trigger'),
  }
}

/**
 * Provide an explanation or instructions for a form field that most users do not need.
 *
 * @version ./package.json
 */
class Tooltip extends React.Component {
  state = {
    open: false,
  }

  componentDidUpdate() {
    if (this.state.open) {
      document.addEventListener('click', this.toggleBubbleOnOutsideEvent)
      document.addEventListener('keypress', this.toggleBubbleOnOutsideEvent)
    } else {
      document.removeEventListener('click', this.toggleBubbleOnOutsideEvent)
      document.removeEventListener('keypress', this.toggleBubbleOnOutsideEvent)
    }
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.toggleBubbleOnOutsideEvent)
    document.removeEventListener('keypress', this.toggleBubbleOnOutsideEvent)
  }

  toggleBubbleOnOutsideEvent = event => {
    const { connectedFieldLabel } = this.props

    const { bubbleId, triggerId } = getIds(connectedFieldLabel)

    const inBubble = closest(event.target, `#${bubbleId}`)
    const inTrigger = closest(event.target, `#${triggerId}`)

    if (!inBubble && !inTrigger) {
      this.toggleBubble()
    }
  }

  toggleBubble = () => {
    this.setState(({ open }) => {
      return { open: !open }
    })
  }

  render() {
    const { direction, connectedFieldLabel, children, ...rest } = this.props

    const { bubbleId, triggerId } = getIds(connectedFieldLabel)

    return (
      <div {...safeRest(rest)} className={iconWrapperStyles.fixLineHeight}>
        <StandaloneIcon
          symbol="questionMarkCircle"
          a11yText={getTriggerA11yText(connectedFieldLabel)}
          onClick={this.toggleBubble}
          id={triggerId}
          aria-controls={bubbleId}
          aria-haspopup="true"
          aria-expanded={this.state.open ? 'true' : 'false'}
        />
        <Responsive
          defaultMatches
          maxWidth="sm"
          render={() => (
            <Bubble id={bubbleId} direction="left" width="full" open={this.state.open}>
              {children}
            </Bubble>
          )}
        />
        <Responsive
          defaultMatches={false}
          minWidth="sm"
          maxWidth="md"
          render={() => (
            <Bubble id={bubbleId} direction="left" width="half" open={this.state.open}>
              {children}
            </Bubble>
          )}
        />
        <Responsive
          defaultMatches={false}
          minWidth="md"
          render={() => (
            <Bubble id={bubbleId} direction={direction} width="quarter" open={this.state.open}>
              {children}
            </Bubble>
          )}
        />
      </div>
    )
  }
}

Tooltip.propTypes = {
  /**
   * Open the bubble to the left or right of the trigger.
   */
  direction: PropTypes.oneOf(['left', 'right']),
  /**
   * The input field that using this tooltip must pass its label so that the tooltip trigger can connect itself. Do not
   * show this prop as it is not part of the public API. If this prop is not specified, the tooltip will generate
   * a generic ID, which could produce duplicate ids.
   *
   * @ignore
   */
  connectedFieldLabel: PropTypes.string,
  /**
   * The message. Can be raw text or text components.
   */
  children: PropTypes.node.isRequired,
}

Tooltip.defaultProps = {
  direction: 'right',
  connectedFieldLabel: undefined,
}

export default Tooltip
