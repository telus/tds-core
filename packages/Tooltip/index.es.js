import Tooltip from './dist/index.es'

export default Tooltip
