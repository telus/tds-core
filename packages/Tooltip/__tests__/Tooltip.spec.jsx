import React from 'react'
import { mount } from 'enzyme'

import StandaloneIcon from '@tds/core-standalone-icon'
import Text from '@tds/core-text'

import Tooltip from '../Tooltip'

const mockMath = Object.create(global.Math)
mockMath.random = () => 0.5
global.Math = mockMath

describe('Tooltip', () => {
  const defaultChildren = 'Tooltip text'

  const doMount = (overrides = {}, children = defaultChildren, options = {}) => {
    const tooltip = mount(<Tooltip {...overrides}>{children}</Tooltip>, options)

    const findTrigger = () => tooltip.find(StandaloneIcon)

    return {
      tooltip,
      findTrigger,
      findBubble: () => tooltip.find('div[data-testid="bubble"]'),
      toggleBubble: () => findTrigger().simulate('click'),
    }
  }

  it('renders', () => {
    const { tooltip } = doMount()

    expect(tooltip).toMatchSnapshot()
  })

  it('has a trigger', () => {
    const { findTrigger } = doMount()

    expect(findTrigger()).toHaveProp('symbol', 'questionMarkCircle')
    expect(findTrigger()).toHaveProp('a11yText', 'Reveal additional information.')
  })

  it('has small text in the bubble', () => {
    const { toggleBubble, findBubble } = doMount({}, 'Some content')
    toggleBubble()

    expect(findBubble()).toContainReact(<Text size="small">Some content</Text>)
  })

  describe('interactivity', () => {
    const createRootElement = () => {
      const root = document.createElement('div')
      root.setAttribute('id', `tooltip-${Math.random()}`)

      return root
    }

    // Not sure why `expect(findBubble()).toHaveClassName(...)` doesn't work
    const getBubbleClassNames = tooltip =>
      tooltip
        .find('[data-testid="bubble"]')
        .at(1)
        .render()
        .attr('class')

    it('shows and hides the bubble', () => {
      const { findBubble, toggleBubble } = doMount()
      expect(findBubble()).toHaveClassName('hide')

      toggleBubble()
      expect(findBubble()).not.toHaveClassName('hide')

      toggleBubble()
      expect(findBubble()).toHaveClassName('hide')
    })

    it('hides the bubble when clicking outside of it', () => {
      const root = createRootElement()

      const { tooltip, toggleBubble } = doMount({}, 'Tooltip text', {
        attachTo: document.body.appendChild(root),
      })

      toggleBubble()
      root.dispatchEvent(new Event('click', { bubbles: true }))

      expect(getBubbleClassNames(tooltip)).toContain('hide')

      // And it removes the event listener so on subsequent clicks it remains hidden
      root.dispatchEvent(new Event('click'), { bubbles: true })

      expect(getBubbleClassNames(tooltip)).toContain('hide')
    })

    it('will not hide the bubble when clicking inside the bubble', () => {
      const root = createRootElement()

      const { tooltip, toggleBubble } = doMount({}, 'Tooltip text', {
        attachTo: document.body.appendChild(root),
      })

      toggleBubble()
      root
        .querySelector('[data-testid="bubble"]')
        .dispatchEvent(new Event('click', { bubbles: true }))

      expect(getBubbleClassNames(tooltip)).not.toContain('hide')
    })
  })

  describe('accessibility', () => {
    it('connects the bubble message to the trigger button for screen readers', () => {
      const { findTrigger, findBubble } = doMount({ connectedFieldLabel: 'Some field' })

      expect(findBubble()).toHaveProp('id', 'some-field_tooltip')
      expect(findBubble()).toHaveProp('role', 'tooltip')
      expect(findBubble()).toHaveProp('aria-live', 'assertive')

      expect(findTrigger()).toHaveProp('aria-haspopup', 'true')
      expect(findTrigger()).toHaveProp('aria-controls', 'some-field_tooltip')
    })

    it('shows and hides the bubble for screen readers', () => {
      const { findTrigger, findBubble, toggleBubble } = doMount()

      expect(findTrigger()).toHaveProp('aria-expanded', 'false')
      expect(findBubble()).toHaveProp('aria-hidden', 'true')

      toggleBubble()

      expect(findTrigger()).toHaveProp('aria-expanded', 'true')
      expect(findBubble()).toHaveProp('aria-hidden', 'false')
    })
  })

  it('passes additional attributes to the wrapper element', () => {
    const { tooltip } = doMount({ 'data-some-attr': 'some value' })

    expect(tooltip).toHaveProp('data-some-attr', 'some value')
  })

  it('does not allow custom CSS', () => {
    const { tooltip } = doMount({
      className: 'some-custom-class',
      style: { color: 'hotpink' },
    })

    const tt = tooltip.find('div[data-testid="tooltipContainer"]')

    expect(tt).not.toHaveProp('className', 'some-custom-class')
    expect(tt).not.toHaveProp('style')
  })

  it('has a default when not connected to any form field', () => {
    const { tooltip } = doMount()

    expect(tooltip.find(StandaloneIcon)).toHaveProp('a11yText', 'Reveal additional information.')
  })
})
