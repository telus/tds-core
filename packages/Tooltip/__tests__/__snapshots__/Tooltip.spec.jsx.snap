// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders 1`] = `
<Tooltip
  direction="auto"
>
  <div
    className="fixLineHeight tooltip"
  >
    <div
      className="tooltipContainer"
      data-testid="tooltipContainer"
    >
      <Bubble
        direction="right"
        id="standalone-tooltip-mc41_tooltip"
        open={false}
        width={
          Object {
            "width": "calc(1008px - 1rem - 0.5rem)",
          }
        }
      >
        <Box
          aria-hidden="true"
          aria-live="assertive"
          dangerouslyAddClassName="right hide"
          data-testid="bubble"
          horizontal={3}
          id="standalone-tooltip-mc41_tooltip"
          inline={false}
          role="tooltip"
          tag="div"
          vertical={2}
        >
          <div
            aria-hidden="true"
            aria-live="assertive"
            className="TDS_Box-modules__horizontalPadding-3___2uoUp TDS_Box-modules__verticalPadding-2___1Uh2T right hide"
            data-testid="bubble"
            id="standalone-tooltip-mc41_tooltip"
            role="tooltip"
          >
            <div
              className="innerBubble"
              style={
                Object {
                  "width": "calc(1008px - 1rem - 0.5rem)",
                }
              }
            >
              <Text
                block={false}
                bold={false}
                invert={false}
                size="small"
              >
                <span
                  className="TDS_Typography-modules__small___mt5lS TDS_Typography-modules__wordBreak___3OZx_ TDS_Typography-modules__smallFont___3S6nz TDS_Typography-modules__color___2CNH8"
                >
                  Tooltip text
                </span>
              </Text>
            </div>
          </div>
        </Box>
      </Bubble>
      <StandaloneIcon
        a11yText="Reveal additional information."
        aria-controls="standalone-tooltip-mc41_tooltip"
        aria-expanded="false"
        aria-haspopup="true"
        id="standalone-tooltip-mc41_trigger"
        onClick={[Function]}
        size={24}
        symbol="questionMarkCircle"
      >
        <Clickable
          aria-controls="standalone-tooltip-mc41_tooltip"
          aria-expanded="false"
          aria-haspopup="true"
          dangerouslyAddStyle={
            Object {
              "margin": "-4px",
              "padding": "4px",
            }
          }
          id="standalone-tooltip-mc41_trigger"
          onClick={[Function]}
          type="button"
        >
          <button
            aria-controls="standalone-tooltip-mc41_tooltip"
            aria-expanded="false"
            aria-haspopup="true"
            className="TDS_Clickable-modules__clickable___Wf5Qr TDS_Spacing-modules__noSpacing___XPYDG TDS_Borders-modules__none___1fCjZ TDS_Forms-modules__font___g7OGX TDS_Typography-modules__color___2CNH8"
            id="standalone-tooltip-mc41_trigger"
            onClick={[Function]}
            style={
              Object {
                "margin": "-4px",
                "padding": "4px",
              }
            }
            type="button"
          >
            <Icon
              aria-label="Reveal additional information."
              size={24}
              symbol="questionMarkCircle"
            >
              <i
                aria-label="Reveal additional information."
                className="TDS_Icon-modules__iconQuestionMarkCircle___2km-5 TDS_Icon-modules__icon___13xYd TDS_Icon-modules__size24___37pQ7"
              />
            </Icon>
          </button>
        </Clickable>
      </StandaloneIcon>
    </div>
  </div>
</Tooltip>
`;
